cluster/restore-from-backup.sh

```diff
@@ -0,0 +1,188 @@
+  for i in $(seq 120); do
+    etcd=$(docker ps | grep etcd | grep -v etcd-empty-dir | wc -l)
+    apiserver=$(docker ps | grep apiserver | wc -l)
+    # TODO: Theoretically it is possible, that apiserver and or etcd
+    # are currently down, but Kubelet is now restarting them and they
+    # will reappear again. We should avoid it.
+    if [ "${etcd}" -eq "0" -a "${apiserver}" -eq "0" ]; then
+      return 0
+    fi
+    sleep 1
+  done
+  return 1
+}
+
+# Move the manifest files to stop etcd and kube-apiserver
+# while we swap the data out from under them.
+mv /etc/kubernetes/manifests/kube-apiserver.manifest ./
+mv /etc/kubernetes/manifests/etcd.manifest ./
+mv /etc/kubernetes/manifests/etcd-events.manifest ./
+
+# Wait for the pods to be stopped
+echo "Waiting for etcd and kube-apiserver to be down"
+if ! wait_for_etcd_and_apiserver_down; then
+  # Couldn't kill etcd and apiserver.
+  echo "Downing etcd and apiserver failed"
+  exit 1
+fi
+
+# Create the sort of directory structure that etcd expects.
+BACKUP_DIR="/var/tmp/backup"
+if [ "${ETCD_API}" == "etcd2" ]; then
+  echo "Preparing etcd backup data for restore"
+  # In v2 mode, we simply copy both snap and wal files to a newly created
+  # directory. After that, we start etcd with --force-new-cluster option
+  # that (according to the etcd documentation) is required to recover from
+  # a backup.
+  mkdir -p "${BACKUP_DIR}/member/snap"
+  mkdir -p "${BACKUP_DIR}/member/wal"
+  # If the cluster is relatively new, there can be no .snap file.
+  mv *.snap "${BACKUP_DIR}/member/snap/" || true
+  mv *.wal "${BACKUP_DIR}/member/wal/"
+
+  # TODO(jsz): This won't work with HA setups (e.g. do we need to set --name flag)?
+  image=$(docker run -d -v ${BACKUP_DIR}:/var/etcd/data \
+    "gcr.io/google_containers/etcd:${ETCD_VERSION}" /bin/sh -c \
+    "/usr/local/bin/etcd --data-dir /var/etcd/data --force-new-cluster")
+
+  # FIXME: Wait until this etcd is up.
+  sleep 10
+
+  # Kill that etcd instance.
+  docker kill "${image}"
+elif [ "${ETCD_API}" == "etcd3" ]; then
+  echo "Preparing etcd snapshot for restore"
+  mkdir -p "${BACKUP_DIR}"
+  mv *.db "${BACKUP_DIR}/"
+  # FIXME: Ensure that there is exactly one file.
+  snapshot="$(ls ${BACKUP_DIR})"
+
+  # Run etcdctl snapshot restore command and wait until it is finished.
+  # setting with --name in the etcd manifest file and then it seems to work.
+  # TODO(jsz): This command may not work in case of HA.
+  image=$(docker run -d -v ${BACKUP_DIR}:/var/tmp/backup --env ETCDCTL_API=3 \
+    "gcr.io/google_containers/etcd:${ETCD_VERSION}" /bin/sh -c \
+    "/usr/local/bin/etcdctl snapshot restore ${BACKUP_DIR}/${snapshot} --name ${NAME} --initial-cluster ${NAME}=http://localhost:2380; mv /${NAME}.etcd/member /var/tmp/backup/")
+  echo "Prepare container exit code: $(docker wait ${image})"
+
+  rm -f "${BACKUP_DIR}/${snapshot}"
+fi
+# Also copy version.txt file
+cp "${VERSION_FILE}" "${BACKUP_DIR}"
+
+# Find out if we are running GCI vs CVM
```

Comment:
Since this seems to be GCP-specific, would it make sense to stick it into cluster/gce instead of at the root? Otherwise, we should make it easier to parameterize this type of data to make it work on other cloud providers.